packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/math_prop.rs

```diff
@@ -1 +1,385 @@
+use crate::{
+    components::prelude::*,
+    state::types::math_expr::{MathExpr, MathOrPrimitive, MathParser},
+};
 
+use super::util::MathOrString;
+
+#[derive(Debug, Default)]
+pub struct MathProp {
+    /// The data query that indicates how the dependencies of this state variable will be created.
+    data_query: DataQuery,
+
+    default_value: MathExpr,
+
+    // TODO: this should be based on a data query for a prop/attribute once we implement enum props or attributes
+    /// A enum determining whether we should use the latex or text parser.
+    parser: MathParser,
+
+    /// Data query that should return a `PropView<bool>`.
+    ///
+    /// If `true`, we split multi-character symbols into the product of the characters
+    /// when parsing the string to a math expression
+    split_symbols: DataQuery,
+
+    // TODO: this should be based on a data query for a prop/attribute once we implement array props or attributes
+    /// A vector of the symbols that should be treated as functions if they are followed by parentheses.  
+    function_symbols: Vec<String>,
+
+    /// A cached value of the expression template used to calculate the final mathematical expression,
+    /// saved here in order to prevent the need for its recalculation if only math values change
+    expression_template_cache: Option<MathExpr>,
```

Comment:
Perhaps this is why you needed things to be `mut`? If we're going to have various bits that need caching on different components, perhaps we should set up a general caching layer in Core. A cache could be returned along with the `DataQuery` results. That way the functions could focus on being pure functions instead of worrying about caching stuff. (Caching will also make `cloning` a bigger operation)

Reply:
This isn't the bad reason for all the `mut`. (It's the tracking of what's changed that is the really annoying part that should be removed with refactoring.)

We do have a caching layer between this and Core that caches the `RequiredData`, which I toyed with extending for this case. But, it seemed like too heavy of a lift for this first pass, as it would require macros to deal with keeping everything properly typed.

I feel like I don't yet have a clear enough picture of how this should work to implement the caching. I'm going to have to revisit all this when implementing `invert` for math, the complexity of which will shed more light on needed functionality.